File: src/sensors/rg15.c
#include <stdio.h>
#include <string.h>
#include <fcntl.h>
#include <termios.h>
#include <unistd.h>
#include <stdlib.h>
#include "rg15.h"

#define MAX_BUFFER 1024
#define DEFAULT_DEVICE "/dev/ttyUSB0"
#define DARWIN_DEVICE "/dev/tty.usbserial-0001"

typedef struct {
    int fd;         // File descriptor
    char device[64];
    struct termios tty;
} RG15Device;

// Initialize the 
RG15Device* rg15_init(const char* device) {
    RG15Device* dev = malloc(sizeof(RG15Device));
    if (!dev) {
        perror("Failed to allocate device structure");
        return NULL;
    }

    // Select appropriate device based on platform
    #ifdef __APPLE__
        strncpy(dev->device, DARWIN_DEVICE, sizeof(dev->device) - 1);
    #else
        strncpy(dev->device, device ? device : DEFAULT_DEVICE, sizeof(dev->device) - 1);
    #endif

    printf("Opening device: %s\n", dev->device);

    // Open serial port
    dev->fd = open(dev->device, O_RDWR |  O_NOCTTY | O_NONBLOCK);
    if (dev->fd < 0) {
        perror("Error opening serial port");
        free(dev);
        return NULL;
    }

    printf("Successfully opened device with fd: %d\n", dev->fd);

    // Configure serial port
    if (tcgetattr(dev->fd, &dev->tty) != 0) {
        perror("Error getting serial port attributes");
        close(dev->fd);
        free(dev);
        return NULL;
    }

    // Set serial port parameters (matching Python's default settings)
    cfsetispeed(&dev->tty, B9600);
    cfsetospeed(&dev->tty, B9600);
    dev->tty.c_cflag &= ~PARENB;        // No parity
    dev->tty.c_cflag &= ~CSTOPB;        // 1 stop bit
    dev->tty.c_cflag &= ~CSIZE;
    dev->tty.c_cflag |= CS8;            // 8 bits per byte
    dev->tty.c_cflag &= ~CRTSCTS;       // No hardware flow control
    dev->tty.c_cflag |= CREAD | CLOCAL; // Enable reading & ignore ctrl lines

    dev->tty.c_lflag &= ~ICANON;        // Disable canonical mode
    dev->tty.c_lflag &= ~ECHO;          // Disable echo
    dev->tty.c_lflag &= ~ECHOE;         // Disable erasure
    dev->tty.c_lflag &= ~ECHONL;        // Disable new-line echo
    dev->tty.c_lflag &= ~ISIG;          // Disable interpretation of INTR, QUIT and SUSP

    dev->tty.c_iflag &= ~(IXON | IXOFF | IXANY);   // Turn off software flow control
    dev->tty.c_iflag &= ~(IGNBRK|BRKINT|PARMRK|ISTRIP|INLCR|IGNCR|ICRNL);

    dev->tty.c_oflag &= ~OPOST;         // Prevent special interpretation of output bytes
    dev->tty.c_oflag &= ~ONLCR;         // Prevent conversion of newline to carriage return/line feed

    // Set timeout to 3 seconds (matching Python version)
    dev->tty.c_cc[VTIME] = 30;          // Wait up to 3 seconds (30 deciseconds)
    dev->tty.c_cc[VMIN] = 0;            // No minimum number of characters

    if (tcsetattr(dev->fd, TCSANOW, &dev->tty) != 0) {
        perror("Error setting serial port attributes");
        close(dev->fd);
        free(dev);
        return NULL;
    }

    // Flush anything in the buffer
    tcflush(dev->fd, TCIOFLUSH);
    
    printf("Serial port configured successfully\n");
    return dev;
}
char* rg15_get_data(RG15Device* dev) {
    static char buffer[MAX_BUFFER];
    ssize_t bytes_written;
    
    bytes_written = write(dev->fd, "r\n", 2);
    if (bytes_written != 2) {
        perror("Failed to write command");
        return NULL;
    }
    
    // Small delay to ensure command is sent
    usleep(100000);  // 100ms delay
    
    ssize_t n = 0;
    int retries = 3;
    
    while (retries > 0) {
        n = read(dev->fd, buffer, sizeof(buffer) - 1);
        
        if (n > 0) {
            buffer[n] = '\0';
            char* p = strchr(buffer, '\r');
            if (p) *p = '\0';
            p = strchr(buffer, '\n');
            if (p) *p = '\0';
            
            return buffer;
        }
        
        retries--;
        usleep(100000);  // Wait 100ms between retries
    }
    
    printf("No data received after 3 attempts\n");
    return NULL;
}

void rg15_cleanup(RG15Device* dev) {
    if (dev) {
        close(dev->fd);
        free(dev);
    }
}

char* get_rain() {
    RG15Device* dev = rg15_init(NULL);
    if (!dev) {
        fprintf(stderr, "Failed to initialize device\n");
        return NULL;
    }
    
    char* data = rg15_get_data(dev);
    if (data) {
        rg15_cleanup(dev);
        return data; 
    } else {
        printf("Failed to get data\n");
        return NULL;
    }
    
}

File: src/lib/sensors/rg15/rg15.zig
const std = @import("std");
const os = std.os;
const mem = std.mem;
const fs = std.fs;
const print = std.debug.print;

const MAX_BUFFER = 1024;

pub const RG15Error = error{
    DeviceOpenError,
    AttributeError,
    WriteError,
    ReadError,
    AllocationError,
};

pub const RG15Device = struct {
    fd: fs.File,
    device_path: []const u8,
    termios: std.posix.termios,
    allocator: mem.Allocator,

    pub fn init(allocator: mem.Allocator, device_path: ?[]const u8) !*RG15Device {
        const default_device = if (comptime @import("builtin").target.os.tag == .macos)
            "/dev/cu.usbserial"
        else
            "/dev/ttyUSB0";

        const path = device_path orelse default_device;

        // Allocate the device structure
        var dev = try allocator.create(RG15Device);
        errdefer allocator.destroy(dev);

        // Open the serial port
        dev.fd = try fs.openFileAbsolute(path, .{
            .mode = .read_write,
            .lock = .none,
        });
        errdefer dev.fd.close();

        // Get current terminal attributes
        dev.termios = try std.posix.tcgetattr(dev.fd.handle);
        var tty = dev.termios;

        // Set baud rate
        try std.posix.tcsetattr(dev.fd.handle, .NOW, tty);

        // Configure port settings
        tty.cflag &= ~@as(std.posix.tc_cflag_t, std.posix.PARENB); // No parity
        tty.cflag &= ~@as(os.tcflag_t, os.CSTOPB); // 1 stop bit
        tty.cflag &= ~@as(os.tcflag_t, os.CSIZE); // Clear size bits
        tty.cflag |= os.CS8; // 8 bits per byte
        tty.cflag &= ~@as(os.tcflag_t, os.CRTSCTS); // No hardware flow control
        tty.cflag |= (os.CREAD | os.CLOCAL); // Enable reading & ignore ctrl lines

        // Input flags
        tty.iflag &= ~@as(os.tcflag_t, os.IXON | os.IXOFF | os.IXANY); // No software flow control
        tty.iflag &= ~@as(os.tcflag_t, os.IGNBRK | os.BRKINT | os.PARMRK | os.ISTRIP | os.INLCR | os.IGNCR | os.ICRNL);

        // Output flags
        tty.oflag &= ~@as(os.tcflag_t, os.OPOST); // Raw output
        tty.oflag &= ~@as(os.tcflag_t, os.ONLCR); // No CR/LF conversion

        // Local flags
        tty.lflag &= ~@as(os.tcflag_t, os.ICANON | os.ECHO | os.ECHOE | os.ECHONL | os.ISIG);

        // Set timeouts
        tty.cc[os.V.TIME] = 30; // 3 seconds timeout (30 deciseconds)
        tty.cc[os.V.MIN] = 0; // No minimum characters

        // Apply settings
        try std.posix.tcsetattr(dev.fd.handle, .NOW, tty);

        // Flush buffers
        try std.posix.tcflush(dev.fd.handle, os.TCIOFLUSH);

        dev.device_path = path;
        dev.allocator = allocator;

        return dev;
    }

    pub fn deinit(self: *RG15Device) void {
        self.fd.close();
        self.allocator.destroy(self);
    }

    pub fn getData(self: *RG15Device) !?[]const u8 {
        var buffer: [MAX_BUFFER]u8 = undefined;

        // Write command
        const cmd = "r\n";
        _ = try self.fd.write(cmd);

        // Small delay
        std.time.sleep(100 * std.time.ns_per_ms); // 100ms delay

        var retries: u8 = 3;
        while (retries > 0) : (retries -= 1) {
            const bytes_read = try self.fd.read(&buffer);

            if (bytes_read > 0) {
                var data = buffer[0..bytes_read];

                // Remove CR/LF
                if (mem.indexOf(u8, data, "\r")) |cr_pos| {
                    data = data[0..cr_pos];
                }
                if (mem.indexOf(u8, data, "\n")) |lf_pos| {
                    data = data[0..lf_pos];
                }

                return data;
            }

            std.time.sleep(100 * std.time.ns_per_ms); // Wait 100ms between retries
        }

        print("No data received after 3 attempts\n", .{});
        return null;
    }
};

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    var dev = try RG15Device.init(allocator, null);
    defer dev.deinit();

    if (try dev.getData()) |data| {
        print("Received data: {s}\n", .{data});
    } else {
        print("No data received\n", .{});
    }
}
